db/snapshot/backup_task.cc

```diff
@@ -48,49 +48,57 @@ tasks::is_abortable backup_task_impl::is_abortable() const noexcept {
     return tasks::is_abortable::yes;
 }
 
-void backup_task_impl::do_backup() {
-    if (!file_exists(_snapshot_dir.native()).get()) {
+future<> backup_task_impl::do_backup() {
+    if (!co_await file_exists(_snapshot_dir.native())) {
         throw std::invalid_argument(fmt::format("snapshot does not exist at {}", _snapshot_dir.native()));
     }
 
     gate uploads;
-    auto wait = defer([&uploads] { uploads.close().get(); });
-
     auto snapshot_dir_lister = directory_lister(_snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
-    auto close_snapshot_dir_lister = deferred_close(snapshot_dir_lister);
 
-    while (auto component_ent = snapshot_dir_lister.get().get()) {
+    std::exception_ptr ex;
+    while (auto component_ent = co_await snapshot_dir_lister.get()) {
```

Comment:
If snapshot_dir_lister.get() throws the snapshot_dir_lister remains non-closed

Reply:
thanks for pointing this out. will fix this.